l2geth/core/vm/evm.go

```diff
@@ -32,10 +34,54 @@ import (
 	"github.com/ethereum/go-ethereum/rollup/dump"
 )
 
+// codec is a decoder for the return values of the execution manager. It decodes
+// (bool, bytes) from the bytes that are returned from both
+// `ExecutionManager.run()` and `ExecutionManager.simulateMessage()`
+var codec abi.ABI
+
+// innerData represents the results returned from the ExecutionManager
+// that are wrapped in `bytes`
+type innerData struct {
+	Success    bool   `abi:"_success"`
+	ReturnData []byte `abi:"_returndata"`
+}
+
+// runReturnData represents the actual return data of the ExecutionManager.
+// It wraps (bool, bytes) in an ABI encoded bytes
+type runReturnData struct {
+	ReturnData []byte `abi:"_returndata"`
+}
+
 // Will be removed when we update EM to return data in `run`.
 var deadPrefix, fortyTwoPrefix, zeroPrefix []byte
 
 func init() {
+	const abidata = `
+	[
+		{
+			"type": "function",
```

Comment:
Rename this to something canonical

Reply:
@gakonst I'd like to define a common interface that defines the interaction between system contracts and userland contracts. This ABI would need to decode 2 things. One method decodes `bytes` and another method decodes `bool,bytes`.

Methods that need to use the interface:
- `ExecutionManager.run`
- `ExecutionManager.simulateMessage`
- `OVM_L2CrossDomainMessenger.relayMessage`

If this interface is followed, then geth will index the transactions based on the userland contracts instead of the system contracts